From b148a73497ac6531029c261584c5affdfda0b492 Mon Sep 17 00:00:00 2001
From: reshmic <reshmic@broadcom.com>
Date: Thu, 16 Aug 2012 13:37:20 +0530
Subject: sae: clear crypto stats counter through proc interface

echo 1  to "clear_crypto_stats" to clear the crypto_stats
[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/crypto/sae/nlm_aead.c b/drivers/crypto/sae/nlm_aead.c
index df3c09f..7f6ad2e 100755
--- a/drivers/crypto/sae/nlm_aead.c
+++ b/drivers/crypto/sae/nlm_aead.c
@@ -427,6 +427,7 @@ static int  xlp_3des_setkey(struct crypto_aead *tfm, u8 *key, unsigned int keyle
 		mode, cipher_alg,NLM_CIPHER_MODE_CBC,0,cipher_key,
 		 cipher_keylen, auth_key, auth_keylen);
 	
+	nlm_ctx->stat = TDES_CBC_STAT | h_stat << 8;;
 	memcpy(d_key,&cipher_key[16],8);
         memcpy(&d_key[1],&cipher_key[8],8);
         memcpy(&d_key[2],&cipher_key[0],8);
diff --git a/drivers/crypto/sae/nlm_crypto.c b/drivers/crypto/sae/nlm_crypto.c
index da6ece3..87f2d6e 100644
--- a/drivers/crypto/sae/nlm_crypto.c
+++ b/drivers/crypto/sae/nlm_crypto.c
@@ -652,12 +652,28 @@ nlm_crypto_read_stats_proc(char *page, char **start, off_t off, int count,
         return len;
 
 }
+int nlm_crypto_clear_stat_proc(struct file *file, const char __user *buffer, 
+		unsigned long count, void *data)
+{
+	char buf[16];
+	unsigned long val;
+
+	copy_from_user(buf, buffer, count);
+	val =  simple_strtol(buf, NULL, 0);	
+
+	if ( val  == 1)
+    		reset_crypto_stats();
+
+	return 1;
+
+}
 
 int
 nlm_crypto_init(void)
 {
     int ret = 0;
     struct proc_dir_entry *entry = NULL;
+    struct proc_dir_entry *clear_entry  = NULL;
 
     entry = create_proc_read_entry("crypto_stats", 0, nlm_root_proc,
 		    nlm_crypto_read_stats_proc,
@@ -669,6 +685,12 @@ nlm_crypto_init(void)
 	    ret = -EINVAL;
     }
 
+    clear_entry = create_proc_entry("clear_crypto_stats", S_IFREG|S_IRUGO|S_IWUSR,nlm_root_proc);
+
+    if ( clear_entry != NULL ) {
+	   clear_entry->write_proc = nlm_crypto_clear_stat_proc;
+    }
+
    nlm_hal_get_crypto_vc_nums(&crypto_vc_base, &crypto_vc_limit); 
 
     if (register_xlp_msgring_handler
-- 
1.7.1

